### DISCUSSION TITLE
Secure network boot and serving of the image

### DISCUSSION THREAD

**Discussion by dazzacoy** [ltsp/ltsp#457] [Ideas]
Hi,

I am trying to setup a secure network boot. I have seen your discussion about using HTTP https://github.com/ltsp/ltsp/issues/75. Would this be the right way to go about this as I would think this is is still not encrypted.

Here are the features we currently have
![image](https://user-images.githubusercontent.com/83610641/116980314-fee2b980-acbd-11eb-8357-91abda230d2f.png)

In addition I would like to completely hide the image that is served unless some type of credentials are provided. Is there an alternative to using NFS, such as SSHFS or maybe using NFS through a SSH tunnel? The key would exist on a USB drive

**Comment by alkisg** [comment]
"Secure boot" is a big topic. "Secure network boot" is even bigger. It would involve:

1. Using UEFI in all clients
2. Generating local UEFI signing keys and deploying them to the UEFI firmware of all clients
3. Generating bundles of kernel+initrd and signing them with the keys
4. At THAT point, you can serve a secure kernel + initrd to the clients
5. AFTER you have a secure initrd, then you can get a wireguard key that is either derived from the hardware, or resides in a USB stick
6. Then you can join a secure wireguard network; it will be secure because you will maintain a list of all the private and public keys per computer
7. Then you can use NFS or NBD or HTTP or whatever else you want, it will be secure due to wireguard; see also issue #127

As you can see, very few of these steps are related to LTSP. And most of them would require a lot of effort and research; they certainly won't be solved via a discussion.

Note that the initial rootfs mounting is implemented by initramfs-tools and dracut, NOT by LTSP. After you add wireguard support in the initramfs, and you set up the keys, initramfs-tools will be able to mount the root via e.g. NFS, and **then** LTSP code is called to loop-mount the image; that's after NFS is already mounted. LTSP doesn't mount the NFS, it just uses it.

Initramfs-tools doesn't have SSHFS support, and it wouldn't be a good idea anyway, to mount the whole rootfs over a fuse filesystem.

**Comment by alkisg** [comment]
P.S. 1: I believe that very few PC lab setups in the whole world are "that secure", in **any operating system** or boot method, local or remote.
If a good hacker has unattended physical access in a computer lab, he'll usually be able to at least install some kind of loggers and get the passwords later on.
If on the other hand, you just want "rootfs security" like your initial question is, and you don't care about all the other security sensitive parts, then #127 (wireguard) would be enough for you.

P.S. 2: If you're using USB sticks, you could write the LTSP image there, instead of serving it over the network. After login, SSHFS secures the /home traffic.

**Reply by dazzacoy** [reply]
Thanks, I decided to write the image onto the USB stick and it boots up fine. I am unable to login though.

I seem to have a problem when connecting to the network. I get the following messages on boot

![image](https://user-images.githubusercontent.com/83610641/117199291-ee2b6400-ade1-11eb-9a01-a9e0f569734f.png)

but if I mount with nfs it works fine

All that I have changed is add an extra menu item to ltsp.ipxe which points to the following method 

:disk
set img x86_64
set cmdline_method root=/dev/sda1 ltsp.image=${img}.img loop.max_part=9
goto ltsp

I have also disabled ufw to make sure no rules are blocking it

is this something that you can help me with?

**Comment by maccadia** [comment]
iPXE does have some security features that you can use ([https - imgtrust](https://ipxe.org/crypto) and [login](https://ipxe.org/cmd/login)) but it does not cover the NFS part of LTSP.

**Comment by alkisg** [comment]
@dazzacoy with root=/dev/sda1 initramfs doesn't try to get an IP, so some things like SERVER are missing.
Try something like PRE_INITRD_BOTTOM="(. /scripts/functions && configure_networking)" to force initramfs to initialize networking, and set ltsp.server=ip in cmdline, or SERVER=ip in ltsp.conf.

Unfortunately I don't have enough free time right now to test and tell you all the missing bits for this use case.
